MSC3013: Encrypted Push (#3013)
matrix-org/matrix-doc: proposals/3013-encrypted-push.md

# MSC3013: Encrypted Push

Push notifications have the problem that they typically go through third-party gateways in order to
be delivered, e.g. FCM (Google) or APNs (Apple) and an app-specific gateway (sygnal). In order to
prevent these push gateways from being able to read any sensitive information the `event_id_only` format
was introduced, which only pushes the `event_id` and `room_id` of an event down the push. After
receiving the push message the client can hit the `GET /_matrix/client/r0/rooms/{roomId}/event/{eventId}`
to fetch the full event, and then create the notification based on that.

This, however, introduces the issue of having to perform an additional HTTP request to be able to
display the full event notification. On some systems (e.g. some weird vendor-specific android phones,
or while driving through ~~rural germany~~ places with patchy cellular network availability) this isn't

> always.

Reviewer: This sentence is unfinished?

Author: right, missing "the case"

## Proposal

A new pusher kind `http.encrypted.curve25519-aes-sha2` is introduced. It behaves the same as the pusher
kind `http`, except that the content it pushes is encrypted. That means it inherits all the pusher
data requirements, all the formats etc. It adds a new pusher data field, `public_key`, which is a
(optionally unpadded) base64-encoded curve25519 public key. This new field is not to be added to the
actual push payload being sent to push gateways. As such, setting such a pusher could look as
following:

```
POST /_matrix/client/r0/pushers/set HTTP/1.1
Content-Type: application/json

{
  "lang": "en",
  "kind": "http.encrypted.curve25519-aes-sha2",
  "app_display_name": "Mat Rix",
  "device_display_name": "iPhone 9",
  "profile_tag": "xxyyzz",
  "app_id": "com.example.app.ios",
  "pushkey": "APA91bHPRgkF3JUikC4ENAHEeMrd41Zxv3hVZjC9KtT8OvPVGJ-hQMRKRrZuJAEcl7B338qju59zJMjw2DELjzEvxwYv7hH5Ynpc1ODQ0aT4U4OFEeco8ohsN5PjL1iC2dNtk2BAokeMCg2ZXKqpc8FXKmhX94kIxQ",
  "data": {
    "url": "https://push-gateway.location.here/_matrix/push/v1/notify",

>     "format": "event_id_only",

Reviewer: Not `event_id_only` here? You wouldn't need the `public_key` here?

Author: This is just an example, as the encrypted pusher inherits from http pusher, both `event_id_only` and (with that other msc) `full_event` would work.

Author: The important bits here are `data.public_key` and the different `kind`

Reviewer 2: Yeah, I agree that while this is technically correct, it makes for a slightly confusing example.

Author: The issue is that there are no other speced types yet, we need https://github.com/matrix-org/matrix-doc/pull/2782 for `full_event` :D Should soru just put `full_event` in there anyways?

Reviewer 2: Ah, I see - yeah, it would probably be reasonable to put that and note that the example assumes #2782 is merged

Author: Updated it 👍

  "room_id": "!slw48wfj34rtnrf:example.com",
  "type": "m.room.message",
  "sender": "@exampleuser:matrix.org",
  "sender_display_name": "Major Tom",
  "room_name": "Mission Control",
  "room_alias": "#exampleroom:matrix.org",
  "prio": "high",
  "content": {
    "msgtype": "m.text",
    "body": "I'm floating in a most peculiar way."
  },
  "counts": {
    "unread": 2,
    "missed_calls": 1
  }
}
```

Resulting in the following final message being pushed out to the push gateway:

```json
{
  "notification": {
    "ephemeral": "base64_of_ephemeral_public_key",
    "ciphertext": "base64_of_ciphertext",
    "mac": "base64_of_mac",
    "devices": [
      {
        "app_id": "org.matrix.matrixConsole.ios",
        "pushkey": "V2h5IG9uIGVhcnRoIGRpZCB5b3UgZGVjb2RlIHRoaXM/",
        "pushkey_ts": 12345678,
        "data": {},
        "tweaks": {
          "sound": "bing"
        }
      }
    ]
  }
}
```

## Potential issues

It is currently implied that a homeserver could push the same notification out to multiple devices
at once, by populating the `devices` array with more than one element. Due to the nature of cryptography,
this won't be possible anymore.

> It is still unclear how well this will work with iOS and its limitations, especially concerning badge-only

Reviewer 2: Since iOS can now have an extension to process incoming push, this actually works nicely with new enough iOS.

Author: Yes, but they have to result in a notif. Badge-only push frames are because of an event being marked as read on another client, which only *removes* a notif. Soru is unsure how this works on iOS

Reviewer 2: I'm guessing the way you'd do it would be to move the logic of forming the iOS-specific payload from the push gateway (which can't do it anymore anyway because it can't read the data) into the push extension in the app, so it would both decrypt and then form the appropriate 'aps' payload etc (I assume setting a badge counts as 'displaying a notification').

Author: > I assume setting a badge counts as 'displaying a notification'

This is basically what soru is asking about here.

Additionally, synapse is happy to send out multiple badge count updates with the same count, and then *nothing* would change. Not sure how  that'd work out, either.

These are basically the things soru is asking about in there, rather than assuming it'd work out ^^

Reviewer 3: - When notification is received in iOS, its pre-processed by client's notification extension. decryption is done here.
- Badge count that has to be displayed is processed in the extension.
- Unread count is set by the `unread_count` passed with the notification json.
- Decryption may fail in notification extension for some reasons, so its always consider handling if decryption fails. Now its handled by not encrypting `userInfo ` in the json.

Author: After finally having an iOS development setup soru can see some issues with iOS concerning messages that change the badge only: It currently does not appear that iOS allows the `mutable-content` to modify only the badge of the app, however the `unread` stuff is encrypted with this MSC, meaning that the push gateway is not able to set badge-only messages anymore. An obvious path forward here would be to not encrypt the unread dict, however that would also decrease the security on other platforms. Perhaps making that a separate option, if the `unread` dict should be encrypted, would make sense here.

Author: Okay, so after discussing this issue internally, we have thought of a couple of possible solutions around this. All this affects iOS only:

1. Have a flag to *not* have the homeserver send badge-only updates. Pro: good for privacy. Con: The apps badge only updates after opening the app itself
2. For badge-only messages, include a `is_only_badge` bool in cleartext.
   The push gateway sends the encrypted parts only via the unreliable background push. Pro: good for privacy. Con: Badge updates are unreliable
3. For badge-only messages, include the `counts` dict in cleartext, in addition to the ciphertext.
   The push gateway sends full apns badge updates. Pro: good user experience. Con: Apple is able to see the exact number of unreads you have in real time

Of course all three could be done as well, so that which exact method is picked is up for the implementation. That way android could also pick e.g. (1) just fine and thus iOS isn't really special-cased.

updates if a message was read on another device.

If the gateway does additional processing, like marking call attempts differently, the relevant data
musn't be encrypted.

> 

Reviewer 2: Could be worth noting that as an alternative, any app could prevent Apple/Google from reading content without spec by just having the push gateway encrypt the content before sending it to APNS/GCM. This MSC would mean the push gateway doesn't get the content either though.

Author: done

## Security considerations

> In a first draft symmetric encryption was used. However, using asymmetric encryption seams like the

Reviewer 2: ```suggestion
In a first draft symmetric encryption was used. However, using asymmetric encryption seems like the
```

# MSC3013: Encrypted Push

Push notifications have the problem that they typically go through third-party gateways in order to
be delivered, e.g. FCM (Google) or APNs (Apple) and an app-specific gateway (sygnal). In order to
prevent these push gateways from being able to read any sensitive information the `event_id_only` format
was introduced, which only pushes the `event_id` and `room_id` of an event down the push. After
receiving the push message the client can hit the `GET /_matrix/client/r0/rooms/{roomId}/event/{eventId}`
to fetch the full event, and then create the notification based on that.

This, however, introduces the issue of having to perform an additional HTTP request to be able to
display the full event notification. On some systems (e.g. some weird vendor-specific android phones,
or while driving through ~~rural germany~~ places with patchy cellular network availability) this isn't

> always the case.

Reviewer 4: ```suggestion
always possible.
```

    "format": "full_event",
    "public_key": "GkZgmbbxnYZfFtywxF4K7NUPqA50Kb7TEsyHeVWyHBI"
  },
  "append": false
}
```

Now, when the homeserver pushes out the message, it is to perform the `notification` dict as with the
http pusher, and then encrypt all of its contents, apart from the `devices` key, using the following
algorithm:

1. Generate an ephemeral curve25519 key, and perform an ECDH with the ephemeral key and the public key
   specified when setting the pusher to generate a shared secret. The public half of the ephemeral key,
   encoded using unpadded base64, becomes the `ephemeral` property of the new payload.
2. Using the shared secret, generate 80 bytes by performing an HKDF using SHA-256 as the hash, with
   a salt of 32 bytes of 0, and with the empty string as the info. The first 32 bytes are used as the
   AES key, the next 32 bytes are used as the MAC key, and the last 16 bytes are used as the AES
   initialization vector.
3. Stringify the JSON object, and encrypt it using AES-CBC-256 with PKCS#7 padding. This encrypted
   data, encoded using unpadded base64, becomes the `ciphertext` property of the new payload.

> 4. Pass the raw encrypted data (prior to base64 encoding) through HMAC-SHA-256 using the MAC key

Author: If we include a mac the gateway (e.g. google) could start bruteforcing your private key. They might also be able to bruteforce it easily with other constraints, that e.g. the output has to be valid json. Wanted to ask @uhoreg how much a concern that would be

Reviewer 5: Brute forcing here shouldn't be much of an issue since the key is generated from (I would assume) a cryptographically secure source, so would take a long time to brute force.  It is an issue in cryptographic logins where the key is generated from a password that might not be very secure, but in this context, it should be fine.

Author: Gotcha, thank you for the comment 👍

    "msgtype": "m.text",
    "body": "I'm floating in a most peculiar way."
  },
  "counts": {
    "unread": 2,
    "missed_calls": 1
  }
}
```

Resulting in the following final message being pushed out to the push gateway:

```json
{
  "notification": {
    "ephemeral": "base64_of_ephemeral_public_key",
    "ciphertext": "base64_of_ciphertext",
    "mac": "base64_of_mac",
    "devices": [
      {
        "app_id": "org.matrix.matrixConsole.ios",
        "pushkey": "V2h5IG9uIGVhcnRoIGRpZCB5b3UgZGVjb2RlIHRoaXM/",
        "pushkey_ts": 12345678,
        "data": {},
        "tweaks": {
          "sound": "bing"
        }
      }
    ]
  }
}
```

## Potential issues

It is currently implied that a homeserver could push the same notification out to multiple devices
at once, by populating the `devices` array with more than one element. Due to the nature of cryptography,
this won't be possible anymore.

It is still unclear how well this will work with iOS and its limitations, especially concerning badge-only
updates if a message was read on another device.

If the gateway does additional processing, like marking call attempts differently, the relevant data
musn't be encrypted.

> ## Alternatives

Reviewer 4: Can we not use the same encryption methods used for [Web Push](https://tools.ietf.org/html/rfc8291). This seems to have a number of benefits.

* Libraries already exist to do the encryption and decryption.
  * Some fully integrated like this one for FCM: https://security.googleblog.com/2018/06/end-to-end-encryption-for-push.html
* Avoids double-encrypting when using Web Push anyways (like browser-based clients).
* Uses an open standard that has been heavily reviewed by security experts.

Of course IIUC push notifications don't use Web Push directly anyways so this won't get us all of the way to a standards based push implementation but at least it is half-compatible.

Reviewer 6: The advantage of using web push directly:
- This is E2EE, and gives VAPID
- This is a standard:
    - Libraries are already available
    - Has been reviewed and acknowledged
    - Many webpush to APNS/FCM/HMS gateways already exist
- This can provide push notifications without gateway to
    - Web app and desktop app
    - Android apps using UnifiedPush
- This doesn't have a requirement on the path (making https://github.com/matrix-org/matrix-spec-proposals/pull/2970/files obsolete)

be delivered, e.g. FCM (Google) or APNs (Apple) and an app-specific gateway (sygnal). In order to
prevent these push gateways from being able to read any sensitive information the `event_id_only` format
was introduced, which only pushes the `event_id` and `room_id` of an event down the push. After
receiving the push message the client can hit the `GET /_matrix/client/r0/rooms/{roomId}/event/{eventId}`
to fetch the full event, and then create the notification based on that.

This, however, introduces the issue of having to perform an additional HTTP request to be able to
display the full event notification. On some systems (e.g. some weird vendor-specific android phones,
or while driving through ~~rural germany~~ places with patchy cellular network availability) this isn't
always possible.

This proposal adds a method to encrypt the push message in a way only the recipient client can decrypt
it, allowing the server to send the full event over push again, with this MSC and [MSC2782](https://github.com/matrix-org/matrix-doc/pull/2782).

## Proposal

A new pusher data field, `algorithm`, is introduced for pushers of the kind `http`. It is an enum,

> represented as string. The currently allowed values are `m.plain` and `m.curve25519-aes-sha2`. Is the
> field absent, then an algorithm of `m.plain` is assumed. The algorithms are defined as following:

Reviewer 2: ```suggestion
represented as string. The currently allowed values are `m.plain` and `m.curve25519-aes-sha2`. If the
field is absent, then an algorithm of `m.plain` is assumed. The algorithms are defined as following:
```

algorithm:

1. Generate an ephemeral curve25519 key, and perform an ECDH with the ephemeral key and the public key
   specified when setting the pusher to generate a shared secret. The public half of the ephemeral key,
   encoded using unpadded base64, becomes the `ephemeral` property of the new payload.
2. Using the shared secret, generate 80 bytes by performing an HKDF using SHA-256 as the hash, with
   a salt of 32 bytes of 0, and with the empty string as the info. The first 32 bytes are used as the
   AES key, the next 32 bytes are used as the MAC key, and the last 16 bytes are used as the AES
   initialization vector.
3. Stringify the JSON object, and encrypt it using AES-CBC-256 with PKCS#7 padding. This encrypted
   data, encoded using unpadded base64, becomes the `ciphertext` property of the new payload.
4. Pass the raw encrypted data (prior to base64 encoding) through HMAC-SHA-256 using the MAC key
   generated above. The first 8 bytes of the resulting MAC are base64-encoded, and become the `mac`
   property of the new payload.

This is the same algorithm used currently in the unstable spec for megolm backup, as such it is

> comptible with libolms PkEncryption / PkDecryption methods.

Reviewer 2: Note that we're now moving away from this though, towards symmetric encryption for key backup. It would also be good to have some words on why this encryption scheme was chosen & what cryptographic properties it gives (and doesn't give).

Author: as far as soru is aware the PkEncryption / PkDecryption was also used for asymmetric SSSS, and it seems like a good idea to have some go-to symmetric / asymmetric algorithms to use for the matrix world, though that is obviously beyond the scope of this PR.

Author: If you meant symmetric vs asymmetric cryptography, there is a paragraph on that under "Security considerations"

Author: Another thing to note is that libolms PkEncryption / PkDecryption actually has a bug where it calculates the mac wrong (which soru discovered while working on encrypted push, uhoreg is aware about and said it is ok to talk about publicly as it only affects data integrity and not security itself), so until libolm gets fixed in that regard implementations need to do their own PkDecryption anyways

  "room_id": "!slw48wfj34rtnrf:example.com",
  "type": "m.room.message",
  "sender": "@exampleuser:matrix.org",
  "sender_display_name": "Major Tom",
  "room_name": "Mission Control",
  "room_alias": "#exampleroom:matrix.org",
  "prio": "high",
  "content": {
    "msgtype": "m.text",
    "body": "I'm floating in a most peculiar way."
  },
  "counts": {
    "unread": 2,
    "missed_calls": 1
  }
}
```

Resulting in the following final message being pushed out to the push gateway:

```json
{
  "notification": {
    "ephemeral": "base64_of_ephemeral_public_key",
    "ciphertext": "base64_of_ciphertext",
    "mac": "base64_of_mac",
    "devices": [
      {
        "app_id": "org.matrix.matrixConsole.ios",
        "pushkey": "V2h5IG9uIGVhcnRoIGRpZCB5b3UgZGVjb2RlIHRoaXM/",
        "pushkey_ts": 12345678,
        "data": {
          "algorithm": "m.curve25519-aes-sha2"
        },
        "tweaks": {
          "sound": "bing"
        }
      }
    ]
  }
}
```

## Potential issues

It is currently implied that a homeserver could push the same notification out to multiple devices
at once, by populating the `devices` array with more than one element. Due to the nature of cryptography,

> this won't be possible anymore.

Reviewer 2: I'm not sure this was ever really used anyway, so much as it's a shame that the format now becomes a bit weird, it's probably no great loss. Probably ought to specify above that there should only be a single device in each push for that reason.

Author: Synapse actually hard-codes to only ever have one device in that array, no idea about the other homeservers.

~~It is still unclear how well this will work with iOS and its limitations, especially concerning badge-only

> updates if a message was read on another device.~~ --> This should work?

Reviewer 2: This seems like the kind of question we should have more data on if there's an impl?

Author: This is the current comment thread for that: https://github.com/matrix-org/matrix-doc/pull/3013#discussion_r577625083

> # MSC3013: Encrypted Push

Author: @dbkr 
> but it does slot nicely into the gap for more secure (and/or more efficient) push in non-e2e rooms.

This isn't only good for non-e2ee rooms, but also good for them, as it means that further down the road, building ontop of this, we can easily tackle on more metadata into the push payload. Together with #3014 and a yet un-written MSC to add unread information per room, this would allow to display full push contents without having to perform any additional HTTP request, thus working significantly more reliable in sketchy network areas.

# MSC3013: Encrypted Push

Push notifications have the problem that they typically go through third-party gateways in order to

> be delivered, e.g. FCM (Google) or APNs (Apple) and an app-specific gateway (sygnal). In order to

Reviewer 7: ```suggestion
be delivered, e.g. FCM (Google) or APNs (Apple) and an app-specific gateway (such as [Sygnal](https://github.com/element-hq/sygnal)). In order to
```

prevent these push gateways from being able to read any sensitive information the `event_id_only` format
was introduced, which only pushes the `event_id` and `room_id` of an event down the push. After
receiving the push message the client can hit the `GET /_matrix/client/r0/rooms/{roomId}/event/{eventId}`
to fetch the full event, and then create the notification based on that.

This, however, introduces the issue of having to perform an additional HTTP request to be able to
display the full event notification. On some systems (e.g. some weird vendor-specific android phones,
or while driving through ~~rural germany~~ places with patchy cellular network availability) this isn't
always possible.

This proposal adds a method to encrypt the push message in a way only the recipient client can decrypt

> it, allowing the server to send the full event over push again, with this MSC and [MSC2782](https://github.com/matrix-org/matrix-doc/pull/2782).

Reviewer 8: Does this require MSC2782 or can it work without it? (It is vague to me if there's a dependency -- if there is, I don't think this can be in FCP?)

Author: it can work without. MSC2782 just adds a new `format` to http pushers, which will make push way more useful but requires encrypted push.

## Proposal

> A new pusher data field, `algorithm`, is introduced for pushers of the kind `http`. It is an enum,

Reviewer 9: please could we have some links to the existing spec, so we can see what is changing?

Reviewer 7: ```suggestion
A new [`PusherData`](https://spec.matrix.org/v1.17/client-server-api/#post_matrixclientv3pushersset_request_pusherdata) field, `algorithm`, is introduced for pushers of the kind `http`. It is an enum,
```

represented as string. The currently allowed values are `m.plain` and `m.curve25519-aes-sha2`. If the
field is absent, then an algorithm of `m.plain` is assumed. The algorithms are defined as following:

### `m.plain` algorithm

The `m.plain` algorithm (the default algorithm) denotes that push is to be delivered in plaintext.
That is no change to current http pushers, thus this MSC is backwards compatible.

### `m.curve25519-aes-sha2` algorithm

The `m.curve25519-aes-sha2` algorithm indicates that the push payloads are to be sent encrypted.
For this, two new pusher data fields are added: `public_key` and `counts_only_type`. Both fields are
to be omitted from the actual push payloads being sent to the push gateways.

> The field `public_key` is required. This key is an (optionally unpadded) base64-encoded curve25519

Author: done

 - `none` (default): The unread status is never included in the plaintext payload
 - `boolean`: If the push frame only affects the counts a boolean (`is_counts_only`) denoting
   this is included in the plaintext payload.
 - `full`: If the push frame only affects the counts the full `counts` dict is included in the plaintext
   payload.

As such, setting such a pusher could look as following (assuming
[MSC2782](https://github.com/matrix-org/matrix-doc/pull/2782) has been merged):

```
POST /_matrix/client/r0/pushers/set HTTP/1.1
Content-Type: application/json

{
  "lang": "en",
  "kind": "http",
  "app_display_name": "Mat Rix",
  "device_display_name": "iPhone 9",
  "profile_tag": "xxyyzz",
  "app_id": "com.example.app.ios",
  "pushkey": "APA91bHPRgkF3JUikC4ENAHEeMrd41Zxv3hVZjC9KtT8OvPVGJ-hQMRKRrZuJAEcl7B338qju59zJMjw2DELjzEvxwYv7hH5Ynpc1ODQ0aT4U4OFEeco8ohsN5PjL1iC2dNtk2BAokeMCg2ZXKqpc8FXKmhX94kIxQ",
  "data": {
    "algorithm": "m.curve25519-aes-sha2",
    "url": "https://push-gateway.location.here/_matrix/push/v1/notify",
    "format": "full_event",
    "public_key": "GkZgmbbxnYZfFtywxF4K7NUPqA50Kb7TEsyHeVWyHBI",
    "counts_only_type": "full"
  },
  "append": false
}
```

> Now, when the homeserver pushes out the message, it is to perform the `notification` dict as with the

Reviewer 7: Maybe it's just me but "perform" strikes me as slightly confusing here.

```suggestion
Now, when the homeserver pushes out the message, it is to construct the [`notification`](https://spec.matrix.org/v1.17/push-gateway-api/#post_matrixpushv1notify_request_notification) dict as with the
```

http pusher, and then encrypt all of its contents, apart from the `devices` key, using the following
algorithm:

1. Generate an ephemeral curve25519 key, and perform an ECDH with the ephemeral key and the public key
   specified when setting the pusher to generate a shared secret. The public half of the ephemeral key,
   encoded using unpadded base64, becomes the `ephemeral` property of the new payload.
2. Using the shared secret, generate 80 bytes by performing an HKDF using SHA-256 as the hash, with
   a salt of 32 bytes of 0, and with the empty string as the info. The first 32 bytes are used as the
   AES key, the next 32 bytes are used as the MAC key, and the last 16 bytes are used as the AES
   initialization vector.
3. Stringify the JSON object, and encrypt it using AES-CBC-256 with PKCS#7 padding. This encrypted
   data, encoded using unpadded base64, becomes the `ciphertext` property of the new payload.
4. Pass the raw encrypted data (prior to base64 encoding) through HMAC-SHA-256 using the MAC key
   generated above. The first 8 bytes of the resulting MAC are base64-encoded, and become the `mac`

>    property of the new payload.

Reviewer 10: It's worth mentioning that the server should immediately drop the ephemeral secret key and shared secret after each use (or at least regularly regenerate them). If either secret is leaked then the push content will be decryptable.

Reviewer 7: It feels like this should be in the proposal text as otherwise it possibly won't make it into the spec PR?

>    property of the new payload.

Reviewer 9: could we use AES-GCM or something modern rather than AES-CBC, for improved performance and to save having to make a separate MAC?

(olm/megolm avoided GCM because GCM wasn't terribly widely supported at the time - I don't think that's true any more?)

Reviewer 11: This code was intentionally designed to be similar to the current pk encryption in libolm, just without the bugs in it. But switching to a different cryptographic primitive solves that too, I guess. But it would need some of the crypto experts approval then.

Reviewer 5: The only other (standard) use for libolm's pkencryption is in key backups, and we're moving away from that to symmetric key backup.  So being consistent with other things is probably less of a concern.

AES-GCM should be fine.

Reviewer 12: I would rather argue that the whole lot of `PkEncryption` should not be used.

It has a plethora of problems:

1. It is buggy as mentioned in https://github.com/matrix-org/matrix-spec-proposals/pull/3013#discussion_r766517926.
2. It does not provide support for domain separation.
3. No additional data can be authenticated.
4. It's not a well-defined standard outside of Matrix

On the other hand, nowadays we have HPKE, and [HPKE's one-shot API](https://www.rfc-editor.org/rfc/rfc9180.html#name-single-shot-apis) would fit the need here. In fact it would fit well in the asymmetric room key backup case as well.

          "sound": "bing"
        }
      }
    ]
  }
}
```

## Potential issues

It is currently implied that a homeserver could push the same notification out to multiple devices
at once, by populating the `devices` array with more than one element. Due to the nature of cryptography,
this won't be possible anymore. However, homeserver implementations such as synapse
[1](https://github.com/matrix-org/synapse/blob/d6fb96e056f79de220d8d59429d89a61498e9af3/synapse/push/httppusher.py#L331-L338)
[2](https://github.com/matrix-org/synapse/blob/d6fb96e056f79de220d8d59429d89a61498e9af3/synapse/push/httppusher.py#L357-L365)
[3](https://github.com/matrix-org/synapse/blob/d6fb96e056f79de220d8d59429d89a61498e9af3/synapse/push/httppusher.py#L419-L426)
even hard-code that `devices` array to only contain a single entry, making it unlikely for this flexibility
having been used in the wild.

If the gateway does additional processing, like marking call attempts differently, the relevant data
musn't be encrypted. That means that clients which rely on that can't use this kind of pusher.

### iOS

Sadly iOS is pretty limiting concerning push notifications. While modifying the content of a push frame,

> and thus e2ee push notifications, is possible on iOS, these modified push frames *have* to result in

Reviewer 13: An iOS app can bypass the mandatory display of a user notification for each push by using [com.apple.developer.usernotifications.filtering](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering) entitlement.

The app develop must request it to Apple before using it. Element-iOS uses it now.

a user-visible notification banner. This limitation becomes a problem when only the `counts` dictionary
is updated, so e.g. when a message is being read. For this, the `counts_only_type` setting has been,
so that the push gateway can do additional logic, based on if the notification frame is such a silent

> update.

Reviewer 8: I'm finding the last sentence difficult to read, I think there's a word missing? Could you clarify it.

Author: hmm, no word is missing. what soru was trying to say is that `counts_only_type` stuffs was added to the msc so that push updates that only affect the counts, and do not contain any events, could also work with apns.

Reviewer 8: Should the beginning be:

> For this, the `counts_only_type` setting has been **added to this proposal**, so that the push gateway...

Author: yes, that's much better, thanks
